### PR TITLE
Update perun-node to get depositor feature

### PR DESCRIPTION
**Pull request by anagha-ks**
#### Description
This is to update perun-node to get the depositor feature introduced in go-perun in this [commit.](https://github.com/hyperledger-labs/go-perun/commit/4bede739a3e4ce208390203ddd75aa3e2eb968b3) Updated perun-node accordingly in prior to shared-nonce part2 change as shared-nonce change depends on this change.

##### Category
Improvement task

##### Relevant issue
Relates to [#101](https://github.com/hyperledger-labs/perun-node/issues/101)

#### Testing
All existing tests should pass.

#### Checklist 

- [x] Name is added to the NOTICE file, if it is not present already.
- [x] Changes are rebased onto the target branch.

### CODE REVIEW

File: perun.go [hunk 1]
```diff
@@ -168,10 +168,10 @@ type WireBus interface {
 //
 // It defines methods for deploying contracts; validating deployed contracts and instantiating a funder, adjudicator.
 type ChainBackend interface {
-	DeployAdjudicator() (adjAddr pwallet.Address, _ error)
-	DeployAsset(adjAddr pwallet.Address) (assetAddr pwallet.Address, _ error)
+	DeployAdjudicator(onChainAddr pwallet.Address) (adjAddr pwallet.Address, _ error)
+	DeployAsset(adjAddr, cred pwallet.Address) (assetAddr pwallet.Address, _ error)
 	ValidateContracts(adjAddr, assetAddr pwallet.Address) error
-	NewFunder(assetAddr pwallet.Address) pchannel.Funder
+	NewFunder(assetAddr, addr pwallet.Address) pchannel.Funder
```
Comment (manoranjith) [inline perun.go]: what addr is the second argument, receiverAddr ?

File: perun.go [hunk 2]
```diff
@@ -168,10 +168,10 @@ type WireBus interface {
 //
 // It defines methods for deploying contracts; validating deployed contracts and instantiating a funder, adjudicator.
 type ChainBackend interface {
-	DeployAdjudicator() (adjAddr pwallet.Address, _ error)
-	DeployAsset(adjAddr pwallet.Address) (assetAddr pwallet.Address, _ error)
+	DeployAdjudicator(onChainAddr pwallet.Address) (adjAddr pwallet.Address, _ error)
+	DeployAsset(adjAddr, cred pwallet.Address) (assetAddr pwallet.Address, _ error)
```
Comment (manoranjith) [inline perun.go]: should this be renamed to addr, with a prefix describing what addr it is ?

File: blockchain/ethereum/internal/chain_test.go [hunk 3]
```diff
@@ -36,9 +36,9 @@ func Test_ChainBackend_Deploy(t *testing.T) {
 	rng := rand.New(rand.NewSource(1729))
 	setup := ethereumtest.NewChainBackendSetup(t, rng, 1)
 
-	adjAddr, err := setup.ChainBackend.DeployAdjudicator()
+	adjAddr, err := setup.ChainBackend.DeployAdjudicator(setup.Accs[0].Address())
```
Comment (manoranjith) [inline blockchain/ethereum/internal/chain_test.go]:
```suggestion
        onChainAddr := setup.Accs[0].Address()
   	adjAddr, err := setup.ChainBackend.DeployAdjudicator(onChainAddr)
```

File: blockchain/ethereum/internal/chain_test.go [hunk 4]
```diff
@@ -36,9 +36,9 @@ func Test_ChainBackend_Deploy(t *testing.T) {
...
 	require.NoError(t, err)
-	assetAddr, err := setup.ChainBackend.DeployAsset(adjAddr)
+	assetAddr, err := setup.ChainBackend.DeployAsset(adjAddr, setup.Accs[0].Address())
```
Comment (manoranjith) [inline blockchain/ethereum/internal/chain_test.go]:
```suggestion
	assetAddr, err := setup.ChainBackend.DeployAsset(adjAddr, onChainAddr)
```

File: blockchain/ethereum/internal/chain_test.go [hunk 5]
```diff
@@ -60,16 +60,16 @@ func Test_ChainBackend_ValidateContracts(t *testing.T) {
 func Test_ChainBackend_NewFunder(t *testing.T) {
 	rng := rand.New(rand.NewSource(1729))
 	setup := ethereumtest.NewChainBackendSetup(t, rng, 1)
-	randomAddr1 := ethereumtest.NewRandomAddress(rng)
+	assetAddr := setup.AssetAddr
+	addr := setup.Accs[0].Address()
 
-	assert.NotNil(t, setup.ChainBackend.NewFunder(randomAddr1))
+	assert.NotNil(t, setup.ChainBackend.NewFunder(assetAddr, addr))
```
Comment (manoranjith) [inline blockchain/ethereum/internal/chain_test.go]: Only requirement here is the parameters should be any valid address, need not be onChain or adjAddr. So use randomAddr as before. 
```suggestion
	assert.NotNil(t, setup.ChainBackend.NewFunder(randomAddr1, randomAddr2))
```

File: blockchain/ethereum/internal/chain_test.go [hunk 6]
```diff
@@ -60,16 +60,16 @@ func Test_ChainBackend_ValidateContracts(t *testing.T) {
...
 }
 
 func Test_ChainBackend_NewAdjudicator(t *testing.T) {
 	rng := rand.New(rand.NewSource(1729))
 	setup := ethereumtest.NewChainBackendSetup(t, rng, 1)
-	randomAddr1 := ethereumtest.NewRandomAddress(rng)
-	randomAddr2 := ethereumtest.NewRandomAddress(rng)
 
-	assert.NotNil(t, setup.ChainBackend.NewAdjudicator(randomAddr1, randomAddr2))
```
Comment (manoranjith) [inline blockchain/ethereum/internal/chain_test.go]: same as above, the addresses just need to be valid addresses, so use random addresses as before.

File: blockchain/ethereum/ethereumtest/chain.go [hunk 7]
```diff
@@ -56,9 +56,10 @@ func NewChainBackendSetup(t *testing.T, rng *rand.Rand, numAccs uint) *ChainBack
 	cbEth := newSimContractBackend(t, walletSetup.Accs, walletSetup.Keystore)
 	cb := &internal.ChainBackend{Cb: &cbEth, TxTimeout: ChainTxTimeout}
 
-	adjudicator, err := cb.DeployAdjudicator()
+	addr := walletSetup.Accs[0].Address()
+	adjudicator, err := cb.DeployAdjudicator(addr)
```
Comment (manoranjith) [inline blockchain/ethereum/ethereumtest/chain.go]:
```suggestion
	onChainAddr := walletSetup.Accs[0].Address()
	adjudicator, err := cb.DeployAdjudicator(onChainAddr)
```

File: blockchain/ethereum/ethereumtest/chain.go [hunk 8]
```diff
@@ -56,9 +56,10 @@ func NewChainBackendSetup(t *testing.T, rng *rand.Rand, numAccs uint) *ChainBack
...
 	require.NoError(t, err)
-	asset, err := cb.DeployAsset(adjudicator)
+	asset, err := cb.DeployAsset(adjudicator, addr)
```
Comment (manoranjith) [inline blockchain/ethereum/ethereumtest/chain.go]:
```suggestion
	asset, err := cb.DeployAsset(adjudicator, onChainAddr)
```

File: blockchain/ethereum/internal/chain.go [hunk 9]
```diff
@@ -38,13 +39,19 @@ type ChainBackend struct {
...
 
 // NewFunder initializes and returns an instance of ethereum funder.
-func (cb *ChainBackend) NewFunder(assetAddr pwallet.Address) pchannel.Funder {
-	return pethchannel.NewETHFunder(*cb.Cb, pethwallet.AsEthAddr(assetAddr))
+func (cb *ChainBackend) NewFunder(assetAddr pwallet.Address, cred pwallet.Address) pchannel.Funder {
+	asset := pethwallet.AsWalletAddr(pethwallet.AsEthAddr(assetAddr))
+	acc := accounts.Account{Address: pethwallet.AsEthAddr(cred)}
+	accounts := map[pethchannel.Asset]accounts.Account{*asset: acc}
+	depositors := map[pethchannel.Asset]pethchannel.Depositor{*asset: new(pethchannel.ETHDepositor)}
+	return pethchannel.NewFunder(*cb.Cb, accounts, depositors)
```
Comment (manoranjith) [inline blockchain/ethereum/internal/chain.go]:
```suggestion
func (cb *ChainBackend) NewFunder(assetAddr pwallet.Address, onChainAddr pwallet.Address) pchannel.Funder {
	asset := pethwallet.AsWalletAddr(pethwallet.AsEthAddr(assetAddr))
	onChainAcc := accounts.Account{Address: pethwallet.AsEthAddr(onChainAddr)}
	onChainAccs := map[pethchannel.Asset]accounts.Account{*asset: onChainAcc}
	depositors := map[pethchannel.Asset]pethchannel.Depositor{*asset: new(pethchannel.ETHDepositor)}
	return pethchannel.NewFunder(*cb.Cb, onChainAccs, depositors)
```

File: blockchain/ethereum/internal/chain.go [hunk 10]
```diff
@@ -38,13 +39,19 @@ type ChainBackend struct {
...
 }
 
 // NewAdjudicator initializes and returns an instance of ethereum adjudicator.
-func (cb *ChainBackend) NewAdjudicator(adjAddr, receiverAddr pwallet.Address) pchannel.Adjudicator {
-	return pethchannel.NewAdjudicator(*cb.Cb, pethwallet.AsEthAddr(adjAddr), pethwallet.AsEthAddr(receiverAddr))
+func (cb *ChainBackend) NewAdjudicator(adjAddr, acct pwallet.Address) pchannel.Adjudicator {
+	acc := accounts.Account{Address: pethwallet.AsEthAddr(acct)}
+	return pethchannel.NewAdjudicator(*cb.Cb, pethwallet.AsEthAddr(adjAddr),
+		pethwallet.AsEthAddr(acct), acc)
```
Comment (manoranjith) [inline blockchain/ethereum/internal/chain.go]:
```suggestion
func (cb *ChainBackend) NewAdjudicator(adjAddr, receiverAddr pwallet.Address) pchannel.Adjudicator {
	receiverAcc := accounts.Account{Address: pethwallet.AsEthAddr(receiverAddr)}
	return pethchannel.NewAdjudicator(*cb.Cb, pethwallet.AsEthAddr(adjAddr),
		pethwallet.AsEthAddr(receiverAddr), receiverAcc)
```

File: blockchain/ethereum/internal/chain.go [hunk 11]
```diff
@@ -61,18 +68,21 @@ func (cb *ChainBackend) ValidateContracts(adjAddr, assetAddr pwallet.Address) er
...
 
 // DeployAsset deploys the asset holder contract, setting the adjudicator address to given value.
-func (cb *ChainBackend) DeployAsset(adjAddr pwallet.Address) (pwallet.Address, error) {
+func (cb *ChainBackend) DeployAsset(adjAddr, cred pwallet.Address) (pwallet.Address, error) {
 	ctx, cancel := context.WithTimeout(context.Background(), cb.TxTimeout)
 	defer cancel()
 
-	addr, err := pethchannel.DeployETHAssetholder(ctx, *cb.Cb, pethwallet.AsEthAddr(adjAddr))
+	acc := accounts.Account{Address: pethwallet.AsEthAddr(cred)}
+	addr, err := pethchannel.DeployETHAssetholder(ctx, *cb.Cb, pethwallet.AsEthAddr(adjAddr), acc)
```
Comment (manoranjith) [inline blockchain/ethereum/internal/chain.go]: Is this onChainAcc or receiverAcc ?

**Review by manoranjith — CHANGES_REQUESTED** [review]

File: blockchain/ethereum/internal/chain.go (hunk 11 above)
Comment (anagha-ks) [reply blockchain/ethereum/internal/chain.go]: It is onChainAcc, [renamed](https://github.com/hyperledger-labs/perun-node/pull/114/commits/aebcf94494c5d13ad272ecadd28edfe7b22dc659) the variable.

File: blockchain/ethereum/internal/chain.go (hunk 10 above)
Comment (anagha-ks) [reply blockchain/ethereum/internal/chain.go]: [Renamed](https://github.com/hyperledger-labs/perun-node/pull/114/commits/aebcf94494c5d13ad272ecadd28edfe7b22dc659) variables name.

File: blockchain/ethereum/internal/chain.go (hunk 9 above)
Comment (anagha-ks) [reply blockchain/ethereum/internal/chain.go]: [Renamed](https://github.com/hyperledger-labs/perun-node/pull/114/commits/aebcf94494c5d13ad272ecadd28edfe7b22dc659) variables.

File: blockchain/ethereum/ethereumtest/chain.go (hunk 8 above)
Comment (anagha-ks) [reply blockchain/ethereum/ethereumtest/chain.go]: [Renamed](https://github.com/hyperledger-labs/perun-node/pull/114/commits/aebcf94494c5d13ad272ecadd28edfe7b22dc659) the variable.

File: blockchain/ethereum/ethereumtest/chain.go (hunk 7 above)
Comment (anagha-ks) [reply blockchain/ethereum/ethereumtest/chain.go]: [Renamed](https://github.com/hyperledger-labs/perun-node/pull/114/commits/aebcf94494c5d13ad272ecadd28edfe7b22dc659).

File: blockchain/ethereum/internal/chain_test.go (hunk 5 above)
Comment (anagha-ks) [reply blockchain/ethereum/internal/chain_test.go]: [Done](https://github.com/hyperledger-labs/perun-node/pull/114/commits/aebcf94494c5d13ad272ecadd28edfe7b22dc659).

File: blockchain/ethereum/internal/chain_test.go (hunk 4 above)
Comment (anagha-ks) [reply blockchain/ethereum/internal/chain_test.go]: [Updated](https://github.com/hyperledger-labs/perun-node/pull/114/commits/aebcf94494c5d13ad272ecadd28edfe7b22dc659).

File: blockchain/ethereum/internal/chain_test.go (hunk 3 above)
Comment (anagha-ks) [reply blockchain/ethereum/internal/chain_test.go]: [updated](https://github.com/hyperledger-labs/perun-node/pull/114/commits/aebcf94494c5d13ad272ecadd28edfe7b22dc659).

File: perun.go (hunk 2 above)
Comment (anagha-ks) [reply perun.go]: I have [renamed](https://github.com/hyperledger-labs/perun-node/pull/114/commits/aebcf94494c5d13ad272ecadd28edfe7b22dc659) it as onChainAddr.

File: perun.go (hunk 1 above)
Comment (anagha-ks) [reply perun.go]: It is onChainAddr, I have updated the same in this [commit](https://github.com/hyperledger-labs/perun-node/pull/114/commits/aebcf94494c5d13ad272ecadd28edfe7b22dc659).

File: blockchain/ethereum/internal/chain.go [hunk 12]
```diff
@@ -38,13 +39,19 @@ type ChainBackend struct {
...
+	depositors := map[pethchannel.Asset]pethchannel.Depositor{*asset: new(pethchannel.ETHDepositor)}
+	return pethchannel.NewFunder(*cb.Cb, onChainAccs, depositors)
 }
 
 // NewAdjudicator initializes and returns an instance of ethereum adjudicator.
 func (cb *ChainBackend) NewAdjudicator(adjAddr, receiverAddr pwallet.Address) pchannel.Adjudicator {
-	return pethchannel.NewAdjudicator(*cb.Cb, pethwallet.AsEthAddr(adjAddr), pethwallet.AsEthAddr(receiverAddr))
+	receiverAcc := accounts.Account{Address: pethwallet.AsEthAddr(receiverAddr)}
+	return pethchannel.NewAdjudicator(*cb.Cb, pethwallet.AsEthAddr(adjAddr),
+		pethwallet.AsEthAddr(receiverAddr), receiverAcc)
```
Comment (manoranjith) [inline blockchain/ethereum/internal/chain.go]: As is "onChainAddr" everywhere else, including the interface definition for the same function in perun.go, the same term can be used here as well for consistency,
```suggestion
	onChainAcc := accounts.Account{Address: pethwallet.AsEthAddr(onChainAddr)}
	return pethchannel.NewAdjudicator(*cb.Cb, pethwallet.AsEthAddr(adjAddr),
		pethwallet.AsEthAddr(onChainAddr), onChainAcc)
```

**Review by manoranjith — CHANGES_REQUESTED** [review]

File: blockchain/ethereum/internal/chain.go (hunk 12 above)
Comment (anagha-ks) [reply blockchain/ethereum/internal/chain.go]: [Renamed](https://github.com/hyperledger-labs/perun-node/pull/114/commits/7070f0127dead417d64fc18a13603559ced67452).

**Review by manoranjith — APPROVED** [review]

**Pull request merged**